src/engine/renderer/glsl_source/reliefMapping_fp.glsl

```diff
@@ -43,7 +44,142 @@ vec3 NormalInTangentSpace(vec2 texNormal)
 	vec3 normal;
 
 #if defined(r_normalMapping)
-#if defined(USE_HEIGHTMAP_IN_NORMALMAP)
+#if defined(USE_NORMALMAP_FROM_HEIGHTMAP)
+	// major inspiration was this post by jollyjeffers:
+	//   https://www.gamedev.net/forums/topic/475213-generate-normal-map-from-heightmap-algorithm/#post_4117038
+	// and this Wikipedia article:
+	//   https://en.wikipedia.org/wiki/Sobel_operator
+	//
+	// Various people recommends doing abs() on sample read in case of float image format:
+	//   http://www.catalinzima.com/2008/01/converting-displacement-maps-into-normal-maps/
+	//   https://community.khronos.org/t/heightmap-to-normalmap/58862
+	// Or report issues by not doing it:
+	//   https://gamedev.stackexchange.com/questions/165575/calculating-normal-map-from-height-map-using-sobel-operator
+	// Other people say texture2D clamps it but it's safer to do this.
+
+	// Get texel size
+	vec2 texelSize = 1.0 / vec2(textureSize(u_HeightMap, 0));
+
+#if defined(r_sobelFiltering)
+	// Useful things to know:
+	//
+	// * It's required to get height samples (S) surrounding the current texel (T) to do a sobel filter:
+	//
+	//   S S S
+	//   S T S
+	//   S S S
+	//
+	// * Sobel X kernel is:
+	//
+	//   [ 1  0  -1 ]
+	//   [ 2  0  -2 ]
+	//   [ 1  0  -1 ]
+	//
+	//   See https://en.wikipedia.org/wiki/Sobel_operator#Formulation
+	//
+	// * Sobel Y kernel is:
+	//
+	//   [  1  2  1 ]
+	//   [  0  0  0 ]
+	//   [ -1 -2 -1 ]
+	//
+	//   Which is the rotated sobel X kernel.
+	//
+	// * Tangent space normals are +Z.
+
+	ivec3 offsets;
+	mat3 xCoords;
+	mat3 yCoords;
+	mat3 heights;
+	mat3 sobel;
+
+	// Components will be computed by accumulating values.
+	normal = vec3(0.0, 0.0, 0.0);
+
+	// Set offsets:
+	offsets = ivec3(-1, 0, 1);
+
+	// Set sobel X kernel (beware, line is column with this notation):
+	sobel[0] = vec3( 1.0,  2.0,  1.0);
+	sobel[1] = vec3( 0.0,  0.0,  0.0);
+	sobel[2] = vec3(-1.0, -2.0, -1.0);
+
+	for (int i = 0; i < 3; i++)
+	{
+		for (int j = 0; j < 3; j++)
+		{
+			if (i != 1 && j != 1)
+			{
+				// Compute coordinates:
+				xCoords[i][j] = texNormal.x + texelSize.x * offsets[i];
+				yCoords[i][j] = texNormal.y + texelSize.y * offsets[j];
+
+				// Get surrounding samples:
+				heights[i][j] = abs(texture2D(u_HeightMap, vec2(xCoords[i][j], yCoords[i][j])).r);
+
+				if (i != 1)
+				{
+					// Sobel computation for X component (use the X sobel kernel):
+					normal.x += heights[i][j] * sobel[i][j];
+				}
+
+				if (j != 1)
+				{
+					// Sobel computation for Y component (use the rotated X sobel kernel as Y one):
+					normal.y += heights[i][j] * sobel[j][i];
+				}
+			}
+		}
+	}
+
+	// Reconstruct Z component while making sure to not
+	// take the square root of a negative number.
+	// That may occur because of compression artifacts.
```

Comment:
This comment doesn't make sense here, since you are getting the normals from your calculations, rather than an RGTC texture.

Reply:
For what I understand, the Z component is still reconstructed from X and Y, the later being computed from heightmap.

I'm getting black artifacts with that sand texture at the same place there is pure white pixels that do not look legit in the original heightmap, I'm not sure to understand everything, but I would expect those pure-white pixels artifacts to produce unexpected normal directions.

Reply:
made the comment explicit about this.